test/functional/api/buffer_updates_spec.lua

```diff
@@ -741,3 +742,87 @@ describe('API: buffer events:', function()
   end)
 
 end)
+
+describe('API: buffer events:', function()
+  before_each(function()
+    helpers.clear()
+  end)
+
+  local function lines_subset(first, second)
+    for i = 1,#first do
+      if first[i] ~= second[i] then
+        return false
+      end
+    end
+    return true
+  end
+
+  local function lines_equal(f, s)
+    return lines_subset(f, s) and lines_subset(s, f)
+  end
+
+  local function did_match_somewhere(buffer_lines, expected_lines)
+    local msg = next_msg()
+
+    while(msg ~= nil) do
+      local event = msg[2]
+      if event == "nvim_buf_lines_event" then
+        local args = msg[3]
+        local starts = args[3]
+        local newlines = args[5]
+
+        -- Size of the contained neovim instance is 23 lines, this might change
+        -- with the test setup. Note updates are continguous.
+        assert(#newlines <= 23)
+
+        for i = 1,#newlines do
+          buffer_lines[starts + i] = newlines[i]
+        end
+        -- we don't compare the msg area of the embedded nvim, it's too flakey
+        buffer_lines[23] = nil
+
+        if lines_equal(buffer_lines, expected_lines) then
+          return true
+        end
+      end
+      msg = next_msg()
+    end
+    return false
+  end
+
+  it('terminal with a nested nvim instance', function()
+    local buffer_lines = {}
```

Comment:
AFAICT this could be created in `did_match_somewhere`, no need to pass (and mutate) it as an argument.

Reply:
It's kind of the point to test this exact behavior, namely "externally" (here only by a function-external variable) the buffer state and making sure it would stay correct during consecutive updates. Note that not all updates consist of 23 lines, so to compare our mirrored buffer state we need a notion of "how did it look before the update event".

Am I making sense?

Reply:
Do you mean `expected_lines`? 

`buffer_lines` is set inside `did_match_somewhere`. Only `expected_lines` needs to be passed as an argument.

Reply:
No I meant `buffer_lines`. What would happen if the `nvim_buf_lines_event` only contains 2 lines? We couldn't compare `expected_lines` then, because those contain the full expected buffer. We could of course shrink down `expected_lines` for that comparison, but than we'd be back a step in that we're looking for a `nvim_buf_lines_event` containing some specific event data, which I had problem with (probably due to screen refreshes that weren't fully deterministic between OSes).

Reply:
oh, so later calls to `did_match_somewhere` depend on earlier calls. might want to note that in a comment above `did_match_somewhere`.